// @react-three/drei/core/index.js
import 'blob-polyfill';
import '../helpers/base64.js';
import '../assets/cloud.base64.js';
import '@babel/runtime/helpers/esm/extends';
import '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import 'react';
import 'react-three-fiber';
export { B as Billboard, a2 as Box, d as CameraShake, v as Center, a3 as Circle, C as Cloud, a4 as Cone, a1 as ContactShadows, e as CubeCamera, a as CubicBezierLine, V as CurveModifier, a5 as Cylinder, D as Detailed, f as DeviceOrientationControls, ag as Dodecahedron, b as Effects, E as Environment, ah as Extrude, F as FlyControls, G as GizmoHelper, l as GizmoViewcube, m as GizmoViewport, ae as Icosahedron, ai as Lathe, L as Line, M as MapControls, W as MeshDistortMaterial, X as MeshWobbleMaterial, af as Octahedron, g as OrbitControls, O as OrthographicCamera, aj as Parametric, c as PerspectiveCamera, a7 as Plane, k as PointerLockControls, ad as Polyhedron, P as PositionalAudio, w as Preload, Q as QuadraticBezierLine, R as Reflector, ac as Ring, ak as RoundedBox, al as ScreenQuad, S as Shadow, _ as Sky, a6 as Sphere, a0 as Stars, t as Stats, ab as Tetrahedron, T as Text, a9 as Torus, aa as TorusKnot, h as TrackballControls, j as TransformControls, a8 as Tube, Z as calcPosFromAngles, i as isWebGL2Available, s as meshBounds, Y as shaderMaterial, $ as softShadows, I as useAnimations, x as useAspect, y as useCamera, H as useContextBridge, n as useCubeTexture, z as useDetectGPU, U as useEdgeSplit, J as useFBO, o as useFBX, p as useGLTF, u as useGizmoContext, A as useHelper, am as useMatcapTexture, an as useNormalTexture, q as useProgress, K as useSimplification, N as useTessellation, r as useTexture } from '../useNormalTexture-34932e21.js';
import 'react-merge-refs';
import 'three';
import 'three/examples/jsm/lines/LineGeometry';
import 'three/examples/jsm/lines/LineMaterial';
import 'three/examples/jsm/lines/Line2';
import 'troika-three-text';
import 'three/examples/jsm/loaders/RGBELoader';
import 'use-asset';
import '../helpers/environment-assets.js';
import 'three/examples/jsm/postprocessing/EffectComposer';
import 'three/examples/jsm/postprocessing/RenderPass';
import 'three/examples/jsm/postprocessing/ShaderPass';
import 'three/examples/jsm/shaders/GammaCorrectionShader';
import 'three/examples/jsm/math/SimplexNoise';
import 'three/examples/jsm/controls/DeviceOrientationControls';
import 'three/examples/jsm/controls/FlyControls';
import 'three/examples/jsm/controls/OrbitControls';
import 'three/examples/jsm/controls/TrackballControls';
import 'three/examples/jsm/controls/TransformControls';
import 'lodash.pick';
import 'lodash.omit';
import 'three/examples/jsm/controls/PointerLockControls';
import 'three/examples/jsm/loaders/FBXLoader';
import 'three/examples/jsm/loaders/GLTFLoader';
import 'three/examples/jsm/loaders/DRACOLoader';
import 'three/examples/jsm/libs/meshopt_decoder.module.js';
import 'zustand';
import '@babel/runtime/helpers/esm/defineProperty';
import '../MeshReflectorMaterial-fdc52c9a.js';
import 'three/examples/js/libs/stats.min';
import 'detect-gpu';
import 'three/examples/jsm/modifiers/SimplifyModifier';
import 'three/examples/jsm/modifiers/TessellateModifier';
import 'three/examples/jsm/modifiers/EdgeSplitModifier';
import 'three/examples/jsm/modifiers/CurveModifier';
import 'three/examples/jsm/objects/Sky';
import 'three/examples/jsm/shaders/HorizontalBlurShader';
import 'three/examples/jsm/shaders/VerticalBlurShader';
import '../helpers/matcap-assets.js';
import '../helpers/normal-assets.js';
